src/storage-types/src/read_holds.rs

```diff
@@ -0,0 +1,169 @@
+        // on drop.
+        let other_since = std::mem::take(&mut other.since);
+
+        self.since.extend(other_since.into_iter());
+
+        // Record the new requirements, which we're guaranteed to be possible
+        // because we're only retracing the two merged sinces together with this
+        // in one go.
+        changes.extend(self.since.iter().map(|t| (t.clone(), 1)));
+
+        match self.holds_tx.send((self.id.clone(), changes)) {
+            Ok(_) => (),
+            Err(e) => {
+                panic!("cannot merge ReadHold: {}", e);
+            }
+        }
+    }
+
+    /// Downgrades `self` to the given `frontier`. Returns `Err` when the new
+    /// frontier is `less_than` the frontier at which this [ReadHold] is
+    /// holding.
+    pub fn try_downgrade(&mut self, frontier: Antichain<T>) -> Result<(), anyhow::Error> {
+        if PartialOrder::less_than(&frontier, &self.since) {
+            return Err(anyhow::anyhow!(
+                "new frontier {:?} is beyond current since {:?}",
+                frontier,
+                self.since
+            ));
+        }
+
+        let mut changes = ChangeBatch::new();
+
+        changes.extend(self.since.iter().map(|t| (t.clone(), -1)));
+        changes.extend(frontier.iter().map(|t| (t.clone(), 1)));
+        self.since = frontier;
+
+        if !changes.is_empty() {
+            // If the other side already hung up, that's ok.
+            let _ = self.holds_tx.send((self.id.clone(), changes));
+        }
+
+        Ok(())
+    }
+}
+
+impl<T: TimelyTimestamp> Clone for ReadHold<T> {
+    fn clone(&self) -> Self {
+        if self.id.is_user() {
+            tracing::debug!("cloning ReadHold on {}: {:?}", self.id, self.since);
+        }
+
+        // Let the other end know.
+        let mut changes = ChangeBatch::new();
+
+        changes.extend(self.since.iter().map(|t| (t.clone(), 1)));
+
+        if !changes.is_empty() {
+            // If the other side already hung up, that's ok.
+            match self.holds_tx.send((self.id.clone(), changes)) {
+                Ok(_) => (),
+                Err(e) => {
+                    panic!("cannot clone ReadHold: {}", e);
+                }
+            }
+        }
+
+        Self {
+            id: self.id.clone(),
+            since: self.since.clone(),
+            holds_tx: self.holds_tx.clone(),
+        }
+    }
+}
+
+impl<T: TimelyTimestamp> Drop for ReadHold<T> {
+    fn drop(&mut self) {
+        if !self.since.is_empty() {
+            if self.id.is_user() {
+                tracing::debug!("dropping ReadHold on {}: {:?}", self.id, self.since);
+            }
+            let mut changes = ChangeBatch::new();
+
+            changes.extend(self.since.iter().map(|t| (t.clone(), -1)));
+            self.since.clear();
+
+            if !changes.is_empty() {
+                // If the other side already hung up, that's ok.
+                let _ = self.holds_tx.send((self.id, changes));
```

Comment:
Nit: If you encapsulate the send in a method you don't have to always repeat that hanging up is ok :)

Reply:
ahh, this made me realize that the other comment in `clone()` is wrong. We _do_ care about being able to send. The code was correct, it's just that the comment was outdated.